src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2796,52 +2809,130 @@ private ImmutableArray<Diagnostic> GetDiagnosticsForMethodBodiesInTree(SyntaxTre
+                {
+                    Debug.Assert(reportUnusedUsings);
+
+                    var discarded = new BindingDiagnosticBag(DiagnosticBag.GetInstance());
+                    Debug.Assert(discarded.DiagnosticBag is object);
+
+                    foreach (var otherTree in SyntaxTrees)
+                    {
+                        var trackingSet = UsageOfUsingsRecorededInTrees;
+
+                        if (trackingSet is null)
+                        {
+                            break;
+                        }
+
+                        if (!trackingSet.Contains(otherTree))
+                        {
+                            compileMethodBodiesAndDocComments(filterTree: otherTree, filterSpan: null, discarded, cancellationToken);
+                            registeredUsageOfUsingsInTree(otherTree);
+                            discarded.DiagnosticBag.Clear();
+                        }
+                    }
+
+                    discarded.DiagnosticBag.Free();
+                }
+            }
 
             if (reportUnusedUsings)
             {
                 ReportUnusedImports(tree, bindingDiagnostics, cancellationToken);
             }
 
             return diagnostics.ToReadOnlyAndFree();
+
+            void compileMethodBodiesAndDocComments(SyntaxTree? filterTree, TextSpan? filterSpan, BindingDiagnosticBag bindingDiagnostics, CancellationToken cancellationToken)
+            {
+                MethodCompiler.CompileMethodBodies(
+                                compilation: this,
+                                moduleBeingBuiltOpt: null,
+                                emittingPdb: false,
+                                emitTestCoverageData: false,
+                                hasDeclarationErrors: false,
+                                emitMethodBodies: false,
+                                diagnostics: bindingDiagnostics,
+                                filterOpt: filterTree is object ? (Predicate<Symbol>?)(s => IsDefinedOrImplementedInSourceTree(s, filterTree, filterSpan)) : (Predicate<Symbol>?)null,
+                                cancellationToken: cancellationToken);
+
+                DocumentationCommentCompiler.WriteDocumentationCommentXml(this, null, null, bindingDiagnostics, cancellationToken, filterTree, filterSpan);
+            }
+
+            void registeredUsageOfUsingsInTree(SyntaxTree tree)
+            {
+                var current = UsageOfUsingsRecorededInTrees;
+
+                while (true)
+                {
+                    if (current is null)
+                    {
+                        break;
+                    }
+
+                    var updated = current.Add(tree);
+
+                    if ((object)updated == current)
+                    {
+                        break;
+                    }
+
+                    if (updated.Count == SyntaxTrees.Length)
+                    {
+                        _usageOfUsingsRecorededInTrees = null;
```

Comment:
Does this need to be a `Volatile.Write` to ensure that it can't be reordered on another thread that is doing the interlocked check below?

Reply:
> Does this need to be a Volatile.Write to ensure that it can't be reordered on another thread that is doing the interlocked check below?

I am not sure what is the concern here. Interlocked is guaranteed to operate on the most recent value, I believe.

---
In reply to: [601866669](https://github.com/dotnet/roslyn/pull/52132#discussion_r601866669) [](ancestors = 601866669)

Reply:
Alright. I think I've convinced myself that there's not a problem here.

---
In reply to: [601872182](https://github.com/dotnet/roslyn/pull/52132#discussion_r601872182) [](ancestors = 601872182,601866669)